File: backend/services/recursos_zotero_mapping.py
"""Mapping bidireccional entre columnes de Recursos i camps de Zotero.

Aquest mòdul **no és generat** — és coneixement de Gnosi (les nostres
columnes catalanes) lligat a la nomenclatura oficial de Zotero. Manten-lo
sincronitzat amb `_zotero_item_to_recursos` a `vault_routes.py` (la
mateixa veritat, dos usos: aquell transforma valors, aquest declara la
correspondència per a queries de rellevància).

Per què hi ha llistes (no strings) al valor de `RECURSOS_TO_ZOTERO_FIELDS`:
una columna Recursos pot ser alimentada per diferents camps Zotero segons
el tipus. P. ex. `Llibre/Revista` és:
  - `publicationTitle` en un `journalArticle`
  - `bookTitle`        en un `bookSection`
  - `proceedingsTitle` en un `conferencePaper`
  - `encyclopediaTitle` en un `encyclopediaArticle`

L'ordre dins de cada llista NO és preferent — només defineix el conjunt
de camps Zotero que poden marcar aquesta columna com a "rellevant" per
un tipus donat.
"""
from __future__ import annotations

from backend.services.zotero_schema import ITEM_TYPE_FIELDS

# Columnes canòniques de Recursos (frontmatter del Vault) → camps Zotero
# que poden alimentar-les. Si un camp Recursos no té correspondència
# Zotero (p. ex. `Citation Key` o `Notes`), no apareix aquí.
RECURSOS_TO_ZOTERO_FIELDS: dict[str, list[str]] = {
    'Item Type':       ['itemType'],
    'Title':           ['title'],
    'Authors':         ['creators'],
    'Any':             ['date'],
    'Llibre/Revista':  ['publicationTitle', 'bookTitle', 'proceedingsTitle', 'encyclopediaTitle'],
    'Editorial':       ['publisher'],
    'Lloc':            ['place'],
    'Volum':           ['volume'],
    'Número':          ['issue'],
    'Pàgines':         ['pages'],
    'Edició':          ['edition'],
    'DOI':             ['DOI'],
    'ISBN':            ['ISBN'],
    'ISSN':            ['ISSN'],
    'URL':             ['url'],
    'Idioma':          ['language'],
}

# Invers: zoteroField → primera columna Recursos que el mencioni.
# Útil per a alta manual on volem proposar quina columna del Vault
# omplir per cada camp Zotero del schema.
ZOTERO_FIELD_TO_RECURSOS: dict[str, str] = {
    field: col
    for col, fields in RECURSOS_TO_ZOTERO_FIELDS.items()
    for field in fields
}


def is_field_relevant_for_type(recursos_field: str, zotero_item_type: str) -> bool:
    """True si la columna Recursos pot ser alimentada per qualsevol dels
    camps oficials del `zotero_item_type` donat.

    Si el tipus no existeix al schema o la columna no té correspondència
    Zotero, retorna False (mostra-la a la secció "Altres camps" al modal).
    """
    candidates = RECURSOS_TO_ZOTERO_FIELDS.get(recursos_field)
    if not candidates:
        return False
    type_fields = ITEM_TYPE_FIELDS.get(zotero_item_type, [])
    return any(c in type_fields for c in candidates)

File: frontend/src/components/Vault/recursosZoteroMapping.js
/**
 * Mapping bidireccional entre columnes de Recursos i camps de Zotero.
 *
 * NO és generat — és coneixement de Gnosi (columnes catalanes) lligat a la
 * nomenclatura oficial de Zotero. Mirall exacte de
 * `backend/services/recursos_zotero_mapping.py`. Sempre que es modifiqui
 * un costat, modificar l'altre — els tests de coherència Py↔JS no cobreixen
 * aquest mòdul (no és generat), però el seu drift trencaria L2/L3.
 *
 * Per què hi ha arrays als valors: una columna Recursos pot ser alimentada
 * per diferents camps Zotero segons el tipus. P. ex. `Llibre/Revista` és
 * `publicationTitle` en un `journalArticle` però `bookTitle` en un
 * `bookSection`. L'ordre dins l'array no és preferent.
 */
import { ITEM_TYPE_FIELDS } from './zoteroSchema';

export const RECURSOS_TO_ZOTERO_FIELDS = {
    'Item Type':       ['itemType'],
    'Title':           ['title'],
    'Authors':         ['creators'],
    'Any':             ['date'],
    'Llibre/Revista':  ['publicationTitle', 'bookTitle', 'proceedingsTitle', 'encyclopediaTitle'],
    'Editorial':       ['publisher'],
    'Lloc':            ['place'],
    'Volum':           ['volume'],
    'Número':          ['issue'],
    'Pàgines':         ['pages'],
    'Edició':          ['edition'],
    'DOI':             ['DOI'],
    'ISBN':            ['ISBN'],
    'ISSN':            ['ISSN'],
    'URL':             ['url'],
    'Idioma':          ['language'],
};

// Invers: zoteroField → primera columna Recursos que el mencioni.
export const ZOTERO_FIELD_TO_RECURSOS = Object.fromEntries(
    Object.entries(RECURSOS_TO_ZOTERO_FIELDS).flatMap(
        ([col, fields]) => fields.map(f => [f, col])
    ),
);

/**
 * True si la columna Recursos pot ser alimentada per qualsevol dels camps
 * oficials del `zoteroItemType` donat. False si el tipus no existeix o la
 * columna no té correspondència Zotero (mostra-la a "Altres camps").
 *
 * @param {string} recursosField  — nom de columna del frontmatter (ca-AD)
 * @param {string} zoteroItemType — clau canònica Zotero (p.ex. 'journalArticle')
 * @returns {boolean}
 */
export function isFieldRelevantForType(recursosField, zoteroItemType) {
    const candidates = RECURSOS_TO_ZOTERO_FIELDS[recursosField];
    if (!candidates) return false;
    const typeFields = ITEM_TYPE_FIELDS[zoteroItemType] || [];
    return candidates.some(c => typeFields.includes(c));
}
